# mLOY_Exomes
The repository contains scripts to run the pipeline that estimates mosaic Loss of the Y chromosome from Exome sequencing data.
To run on dnanexus wdl script is available as well.

# Requirements
1. mosdepth tool should be installed (https://github.com/brentp/mosdepth)
2. R
3. tidyr library

# Inputs
1. Bed file with intervals in a genome (exons) in which coverage will be calculated
2. Bed file with the same intervals but with the group for each interval in a separate column. Groups are created after matching exons on their GC content and are used for coverage normalization.
3. Bam/Cram Exome sequencing alignment file with index files.
4. Reference genome (that was used for the alignment) with indexes (.fasta, .fai, and .dict)

# Pipeline
## 1. Calculate the coverage in each exon from bed file
Run 
```
sh run_mosdepth.sh 
```
in your console. Please specify all inputs inside the script (bed file is provided).

## 2. Normalize the coverage of chrY to the autosomes.
It is done by predefined groups by matching exons by GC content. To run this step there is R script. To run:

```
Rscript NormalizeByGCcontentGroups.R prefix.regions.bed.gz exome_groups.bed
```
prefix.regions.bed.gz file is generated by mosdepth on a previous step. Prefix should be unique for each cram or bam file processed (e.g. individual id).

## 3. Aggregate the results

To aggregate results for each processed file into one table run

```
sh aggregate_results.sh > output_file_name

```

## 4. Estimate mosaic loss of the Y chromosome

